compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/types/Types.java

```diff
@@ -0,0 +1,201 @@
+        this.JSON = typesFactory.getTypeDescriptor(symbolTable.jsonType);
+        this.BYTE = typesFactory.getTypeDescriptor(symbolTable.byteType);
+        this.COMPILATION_ERROR = typesFactory.getTypeDescriptor(symbolTable.semanticError);
+    }
+
+    /**
+     * Lookup for the symbol of a user defined type within a given module. This would be considering type
+     * definitions, enums, enum members, and class definitions as valid user defined types when
+     * looking up. The module is determined by the provided org, module name and the version.
+     *
+     * @param org           The organization of the looking up module
+     * @param moduleName    The name of the looking up module
+     * @param version       The version of the looking up module
+     * @param typeDefName   The type definition
+     * @return The {@link Symbol} of the user defined type
+     */
+    public Optional<Symbol> getTypeByName(String org, String moduleName, String version, String typeDefName) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
+
+        return getTypeDefByName(packageID, typeDefName);
+    }
+
+    /**
+     * Lookup for all the symbols of user defined types within a given module. This would be considering type
+     * definitions, enums, enum members, and class definitions as valid user defined types when looking up.
+     * The module is determined by the provided org, module name and the version.
+     *
+     * @param org           The organization of the looking up module
+     * @param moduleName    The name of the looking up module
+     * @param version       The version of the looking up module
+     * @return A {@link Map} of the user defined type symbols
+     */
+    public Optional<Map<String, Symbol>> typesInModule(String org, String moduleName, String version) {
+        PackageID packageID = new PackageID(Names.fromString(org), Names.fromString(moduleName),
+                Names.fromString(version));
+
+        return getTypeDefSymbolsInModule(packageID);
+    }
+
+    public static Types getInstance(CompilerContext context) {
+        Types types = context.get(TYPES_KEY);
+        if (types == null) {
+            types = new Types(context);
+        }
+
+        return types;
+    }
+
+    private Optional<Symbol> getTypeDefByName(PackageID packageID, String typeDefName) {
+        BPackageSymbol packageSymbol = packageCache.getSymbol(packageID);
+        if (packageSymbol == null) {
+            return Optional.empty();
+        }
+
+        SymbolEnv pkgEnv = symbolTable.pkgEnvMap.get(packageSymbol);
+        Scope.ScopeEntry entry = pkgEnv.scope.lookup(Names.fromString(typeDefName));
+        if (entry != null) {
+            if (isValidTypeDef(entry.symbol)) {
+                return Optional.ofNullable(symbolFactory.getBCompiledSymbol(entry.symbol, typeDefName));
```

Comment:
Let's make this `Optional.of()`. No reason the symbol factory to return `null` when we have a valid symbol. If it does, then it's a bug.